util/metric: emit evicted label values in high cardinality metrics

aa-joshi opened the pull request:
Problem:
High cardinality metrics support eviction of child metrics based on configured
policies. This created a discrepancy where the sum of child label values did
not equal the aggregated parent value, since evicted metrics were excluded
from the current snapshot. Users could not reconcile the difference between
the parent metric and the sum of visible child metrics.

Behavior with this change:
This patch introduces a special child metric with `_evicted` as the label value
for all configured labels. The `_evicted` metric represents the sum of all
values that have been evicted and are not part of the current snapshot. Now:
  aggregated value = `_evicted` child metric + sum of current child metrics

The `_evicted` label is emitted for counter and gauge metrics only. We exclude
histograms because an `_evicted` value would distort the distribution.

Why it is helpful:
This change provides full visibility into metric values. Users can now account
for all data: the current active child metrics plus the aggregated evicted
values. This eliminates confusion when the sum of visible children doesn't
match the parent, making it clear that the difference represents evicted data.
Child metric behavior is controlled by `server.child_metrics.enabled`.

Epic: CRDB-53398
Part of: CRDB-57472

Release note (ops change): High cardinality metrics now emit a metric with
label values set to `_evicted` when `server.child_metrics.enabled` is enabled.
This applies to counter and gauge metrics and tracks the aggregated value of
all evicted child metrics, ensuring that the sum of all child values plus the
_evicted value equals the parent aggregated value.

----------

Lets take example of metric `sql.select.count` and how it would behave at `_status/vars` endpoint when it is converted to high cardinality metrics

- `server.child_metrics.enabled: false`
HELP sql_select_count Number of SQL SELECT statements successfully executed
TYPE sql_select_count counter
sql_select_count{node_id="1"} 143

- `server.child_metrics.enabled: true`
HELP sql_select_count Number of SQL SELECT statements successfully executed
TYPE sql_select_count counter
sql_select_count{node_id="1", value_type="evicted"} 1
sql_select_count{node_id="1", app="tpcc", db="tpcc"} 142

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/158662)
<!-- Reviewable:end -->

dhartunian reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158662)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, and @arjunmahishi)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-Of_5pFGE-B3R5hz7Sw-:-Of_5pFGE-B3R5hz7Sw0:bdk0ku6):*
This commit message does not fully explain what's going on here and why you would make this change. Can you explain further how the user is expected to interact with this? The ticket that's referenced does not motivate this change. The phrasing does not clarify the math that's happening here. Does adding up "all the labels" + `evicted` == `aggregated`? 

Additionally, this change emits labels that require customer understanding to interpret correctly and, as far as I know, are non-standard because the "aggregated" value is expected to be computed from the existing labels. This should have a release note, unless we expect to document this elsewhere.

Also, please paste output in text from terminal, not in screenshots so that it's visible in git and when searching.


<!-- Sent from Reviewable.io -->

aa-joshi committed:
util/metric: emit evicted label values in high cardinality metrics

Problem:
High cardinality metrics support eviction of child metrics based on configured
policies. This created a discrepancy where the sum of child label values did
not equal the aggregated parent value, since evicted metrics were excluded
from the current snapshot. Users could not reconcile the difference between
the parent metric and the sum of visible child metrics.

Behavior with this change:
This patch introduces a special child metric with `_evicted` as the label value
for all configured labels. The `_evicted` metric represents the sum of all
values that have been evicted and are not part of the current snapshot. Now:
  aggregated value = `_evicted` child metric + sum of current child metrics

The `_evicted` label is emitted for counter and gauge metrics only. We exclude
histograms because an `_evicted` value would distort the distribution.

Why it is helpful:
This change provides full visibility into metric values. Users can now account
for all data: the current active child metrics plus the aggregated evicted
values. This eliminates confusion when the sum of visible children doesn't
match the parent, making it clear that the difference represents evicted data.
Child metric behavior is controlled by `server.child_metrics.enabled`.

Epic: CRDB-53398
Part of: CRDB-57472

Release note (ops change): High cardinality metrics now emit a metric with
label values set to `_evicted` when `server.child_metrics.enabled` is enabled.
This applies to counter and gauge metrics and tracks the aggregated value of
all evicted child metrics, ensuring that the sum of all child values plus the
_evicted value equals the parent aggregated value.

Abhinav1299 reviewed (commented):
In the PR description its mentioned that `value_type=evicted` is not for histogram metrics, but in the release note section, its other way around? Can you check this once?

aa-joshi commented on pkg/util/metric/aggmetric/histogram.go:
```diff
@@ -487,6 +487,19 @@ func (h *HighCardinalityHistogram) RecordValue(v int64, labelValues ...string) {
+	parentMetric.Label = parentLabels
+	f(parentMetric)
+
+	// Then emit child metrics
```
We don't have to explicitly emit for aggregated as histogram gives you the aggregated value across the buckets.
<!-- Reviewable comment -OfirflLCAP_4Yh-hoAp:bv3dstz -->
<!-- Sent from Reviewable.io -->

aa-joshi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158662)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @arjunmahishi, and @dhartunian)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-Of_5pFGE-B3R5hz7Sw-:-Ofir7Lb6zElrslwMqG2:bneldvo):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This commit message does not fully explain what's going on here and why you would make this change. Can you explain further how the user is expected to interact with this? The ticket that's referenced does not motivate this change. The phrasing does not clarify the math that's happening here. Does adding up "all the labels" + `evicted` == `aggregated`? 

Additionally, this change emits labels that require customer understanding to interpret correctly and, as far as I know, are non-standard because the "aggregated" value is expected to be computed from the existing labels. This should have a release note, unless we expect to document this elsewhere.

Also, please paste output in text from terminal, not in screenshots so that it's visible in git and when searching.
</blockquote></details>

As discussed offline, I have updated the PR.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158662)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, and @arjunmahishi)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-Of_5pFGE-B3R5hz7Sw-:-Ofjns-vEVxMLjWXPQ3B:b72af1x):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

As discussed offline, I have updated the PR.
</blockquote></details>

Sorry to be nitpicky but this paragraph is challenging to read because it's one giant block and contains a mix of motivation, implementation details, and fluff. 

If you don't feel like editing it yourself, here's an example of how ChatGPT can improve on it below. Notably, it splits up the motivation, description of the change, and details into separate scannable sections. All I asked was to make it more "readable".

----

**Problem**  
Aggregated metrics currently emit both a parent value and individual child values with their associated labels. This works when all child metrics are present, but breaks for high-cardinality metrics where child series may be **evicted** based on configuration. When some children have been evicted, the sum of the remaining children no longer matches the parent value, which can confuse readers.

**What this change does**  
This patch introduces an **`evicted`** synthetic child label value. It represents the total value of all child metrics that have been evicted and are no longer part of the current snapshot.

**Resulting behavior**

-   The aggregated value = (`evicted` child value) + (sum of current child values).
    
-   The `evicted` child is emitted for **counter** and **gauge** metrics.
    
-   We **do not** emit an `evicted` child for **histograms**, because it would distort distribution data.
    
-   Whether child metrics exist is controlled by the cluster setting:  
    `server.child_metrics.enabled`.
    

**Why this helps**  
This makes the parent/child relationship understandable again: every part of the parent’s value is accounted for, even when some children have been removed.
___
*[`pkg/util/metric/aggmetric/testdata/HighCardinalityGauge_pre_eviction.txt` line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-OfjqikM5kw0zlbkFRf9:-OfjqikM5kw0zlbkFRfA:b-y1q0sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7592983971b25be061bd0211f4ad4fa0f50c6dcd/pkg/util/metric/aggmetric/testdata/HighCardinalityGauge_pre_eviction.txt#L18)):*
> ```help
> foo_gauge{database="1",application_name="8"} 9
> foo_gauge{database="1",application_name="9"} 10
> foo_gauge{database="evicted",application_name="evicted"} 0
> ```

can we avoid emitting this if there are zero evicted entries? I think in general it might be nice to omit children that are zero.
___
*[`pkg/util/metric/aggmetric/counter.go` line 492 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-OfjqENx4x9HuL_90vJ4:-OfjqENx4x9HuL_90vJ5:bbq12nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7592983971b25be061bd0211f4ad4fa0f50c6dcd/pkg/util/metric/aggmetric/counter.go#L492)):*
> ```Go
> 			Value: &evictedValueLabel,
> 		})
> 	}
> ```

can we store this statically instead of re-computing on every request?
___
*[`pkg/util/metric/aggmetric/counter.go` line 496 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-OfjqN_DEkUC-WuGVbPg:-OfjqN_DEkUC-WuGVbPh:b-bvv1o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7592983971b25be061bd0211f4ad4fa0f50c6dcd/pkg/util/metric/aggmetric/counter.go#L496)):*
> ```Go
> 	evictedMetric := &io_prometheus_client.Metric{
> 		Counter: &io_prometheus_client.Counter{
> 			Value: proto.Float64(float64(c.evictedCounter.Count())),
> ```

Have you considered maintaining the "evicted" counter as just another child? Not sure how possible/difficult this is, but curious. Then this method would remain unchanged.
___
*[`pkg/util/metric/aggmetric/agg_metric.go` line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-Ofjpx7q5O3hKqo9RUzF:-Ofjpx7q5O3hKqo9RUzG:bnc236p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7592983971b25be061bd0211f4ad4fa0f50c6dcd/pkg/util/metric/aggmetric/agg_metric.go#L41)):*
> ```Go
> var (
> 	delimiter         = []byte{'_'}
> 	evictedValueLabel = "evicted"
> ```

is there any way to make this value a bit more "special" like using `_evicted` or something, just so customer-controlled labels are less likely to conflict with it?
___
*[`pkg/util/metric/prometheus_exporter.go` line 162 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158662#-Ofjpr9CE327FaYVfXyP:-Ofjpr9CE327FaYVfXyQ:bz4kz1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7592983971b25be061bd0211f4ad4fa0f50c6dcd/pkg/util/metric/prometheus_exporter.go#L162)):*
> ```Go
> 			}
> 			family.Metric = append(family.Metric, m)
> 			return
> ```

it's unclear why `PrometheusEvictable` can't be handled by the case below. Can you explain?


<!-- Sent from Reviewable.io -->
